dpnp/backend/extensions/statistics/common.hpp

```diff
@@ -0,0 +1,190 @@
+// - Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#pragma once
+
+#include <complex>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+
+// clang-format off
+// math_utils.hpp doesn't include sycl header but uses sycl types
+// so sycl.hpp must be included before math_utils.hpp
+#include <sycl/sycl.hpp>
+#include "utils/math_utils.hpp"
+// clang-format on
+
+namespace statistics
+{
+namespace common
+{
+
+template <typename N, typename D>
+constexpr auto CeilDiv(N n, D d)
+{
+    return (n + d - 1) / d;
+}
+
+template <typename N, typename D>
+constexpr auto Align(N n, D d)
+{
+    return CeilDiv(n, d) * d;
+}
+
+template <typename T, sycl::memory_order Order, sycl::memory_scope Scope>
+struct AtomicOp
+{
+    static void add(T &lhs, const T value)
+    {
+        sycl::atomic_ref<T, Order, Scope> lh(lhs);
+        lh += value;
+    }
+};
+
+template <typename T, sycl::memory_order Order, sycl::memory_scope Scope>
+struct AtomicOp<std::complex<T>, Order, Scope>
+{
+    static void add(std::complex<T> &lhs, const std::complex<T> value)
+    {
+        T *_lhs = reinterpret_cast<T(&)[2]>(lhs);
+        const T *_val = reinterpret_cast<const T(&)[2]>(value);
+        sycl::atomic_ref<T, Order, Scope> lh0(_lhs[0]);
+        lh0 += _val[0];
+        sycl::atomic_ref<T, Order, Scope> lh1(_lhs[1]);
+        lh1 += _val[1];
+    }
+};
+
+template <typename T>
+struct Less
+{
+    bool operator()(const T &lhs, const T &rhs) const
+    {
+        return std::less{}(lhs, rhs);
+    }
+};
+
+template <typename T>
+struct Less<std::complex<T>>
+{
+    bool operator()(const std::complex<T> &lhs,
+                    const std::complex<T> &rhs) const
+    {
+        return dpctl::tensor::math_utils::less_complex(lhs, rhs);
+    }
+};
+
+template <typename T>
+struct IsNan
+{
+    static bool isnan(const T &v)
+    {
+        if constexpr (std::is_floating_point_v<T>) {
```

Comment:
```suggestion
        if constexpr (std::is_floating_point_v<T> || std::is_samve_v<T, sycl::half>) {
```

Reply:
Or I guess `sycl::ext::oneapi::experimental::is_genfloat_v<...>` can be used.

Reply:
Done, thank you.
@antonwolfy while your suggestion looks good, let's for this PR stick to the version proposed by Sasha.